chain/src/main/scala/org/bitcoins/chain/models/BlockHeaderDAO.scala

```diff
@@ -57,21 +58,40 @@ sealed abstract class BlockHeaderDAO
   def getAncestorAtHeight(
       child: BlockHeaderDb,
       height: Long): Future[Option[BlockHeaderDb]] = {
-    //extremely inefficient and probably needs to be re-implemented
-    //we are reading from disk every time here
-    val headerOptF = read(child.previousBlockHashBE)
-    headerOptF.flatMap {
-      case Some(header) =>
-        if (header.height == height) {
-          Future.successful(Some(header))
-        } else if (height <= header.height) {
-          //need to keep traversing backwards to get to the right height
-          getAncestorAtHeight(header, height)
+    val headersF = getBetweenHeights(from = height, to = child.height - 1)
+
+    val headersByHeight: Array[Vector[BlockHeaderDb]] =
+      new Array[Vector[BlockHeaderDb]](_length = (child.height - height).toInt)
+    headersByHeight.indices.foreach(headersByHeight(_) = Vector.empty)
+
+    headersF.map { headers =>
+      headers.foreach { header =>
+        val index = (header.height - height).toInt
+        headersByHeight(index) = headersByHeight(index).:+(header)
+      }
+
+      val groupedByHeightHeaders: Vector[Vector[BlockHeaderDb]] =
+        headersByHeight.toVector
+
+      @tailrec
+      def loop(
+          currentHeader: BlockHeaderDb,
+          headersByDescHeight: Vector[Vector[BlockHeaderDb]]): Option[
+        BlockHeaderDb] = {
+        if (currentHeader.height == height) {
+          Some(currentHeader)
         } else {
-          Future.successful(None)
+          val prevHeaderOpt = headersByDescHeight.headOption.flatMap(
```

Comment:
if you are worried about performance, a List might be faster here http://www.lihaoyi.com/post/BenchmarkingScalaCollections.html#lists-vs-vectors

specifically: 

```


Constructing a List item-by-item is 5-15x (!) faster than constructing a Vector item-by-item.

De-constructing a List item-by-item using .tail is 50-60x (!) faster than de-constructing a Vector item-by-item using .tail

Looking things up by index in a Vector works; looking things up by index in a List is O(n), and expensive for non-trivial lists.
```

Since we aren't looking things up by index here, probably should be using `List`. I guess this will incur more memory, but I think that is ok in this context